### PR TITLE
sql: use a single instance of UnqualifiedStar

### LINKED ISSUE #19159: sql: singleton UnqualifiedStar
Suggested by @nvanbenschoten 

> nit: we should create a single instance of parser.UnqualifiedStar{} and use that everywhere.

### PR DESCRIPTION
**Pull request by knz**
Fixes #19159.

### CODE REVIEW

**Comment by nvanbenschoten** [comment]
I'm confused. I didn't realize we already were using a singleton. We should have just been calling `StarExpr()` in #19150.

**Comment by knz** [comment]
StarExpr() is a VarName. Not suitable for use as element of UnresolvedName.

File: pkg/sql/parser/var_name.go [hunk 1]
```diff
@@ -82,10 +82,7 @@ var _ VarName = &ColumnItem{}
 // NormalizeVarName is a no-op for UnqualifiedStar (already normalized)
 func (u UnqualifiedStar) NormalizeVarName() (VarName, error) { return u, nil }
 
-var singletonStarName VarName = UnqualifiedStar{}
-
-// StarExpr is a convenience function that represents an unqualified "*".
-func StarExpr() VarName { return singletonStarName }
+var UnqualifiedStarInstance = UnqualifiedStar{}
```
Comment (nvanbenschoten) [inline pkg/sql/parser/var_name.go]: Move this to `name_part.go` near the type definition.

**Review by nvb — APPROVED** [review]

File: pkg/sql/parser/var_name.go (hunk 1 above)
Comment (knz) [reply pkg/sql/parser/var_name.go]: Done.

**Comment by petermattis** [comment]
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/parser/name_part.go, line 113 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19160#-Kw684mwWO6FxH5gZz-f:-Kw684mxnM5WORYxKGwl:bbb99zq) ([raw file](https://github.com/cockroachdb/cockroach/blob/74b2140ae2a3e1ba46a5a64c8d1b7571b48cb85a/pkg/sql/parser/name_part.go#L113)):*
> ```Go
> 
> // UnqualifiedStarInstance is the singleton instance of UnqualifiedStar.
> var UnqualifiedStarInstance = UnqualifiedStar{}
> ```

Why is this better than using `UnqualifiedStar{}` everywhere? (I know one reason, but I doubt it is the reason you're using).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19160)*

**Comment by nvanbenschoten** [comment]
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/parser/name_part.go, line 113 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19160#-Kw684mwWO6FxH5gZz-f:-Kw68ujB1Idv_QXvfyx4:b-vq5v9e) ([raw file](https://github.com/cockroachdb/cockroach/blob/74b2140ae2a3e1ba46a5a64c8d1b7571b48cb85a/pkg/sql/parser/name_part.go#L113)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Why is this better than using `UnqualifiedStar{}` everywhere? (I know one reason, but I doubt it is the reason you're using).
</blockquote></details>

I suggested it in a previous review because there's no reason to allocate multiple instances of this. Thinking about it more though, it's zero sized so it shouldn't allocate when put in an interface anyway. Plus, even if it did, Go may [avoid duplicate allocs](https://golang.org/ref/spec#Size_and_alignment_guarantees).

What reason were you thinking? Well-defined comparisons?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19160)*

**Comment by knz** [comment]
merge or no merge?

**Comment by petermattis** [comment]
I'd say no merge, but don't have a strong preference. 

---

Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/name_part.go, line 113 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19160#-Kw684mwWO6FxH5gZz-f:-KwetVLb_HGLTf9hsdLn:b8zuthk) ([raw file](https://github.com/cockroachdb/cockroach/blob/74b2140ae2a3e1ba46a5a64c8d1b7571b48cb85a/pkg/sql/parser/name_part.go#L113)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I suggested it in a previous review because there's no reason to allocate multiple instances of this. Thinking about it more though, it's zero sized so it shouldn't allocate when put in an interface anyway. Plus, even if it did, Go may [avoid duplicate allocs](https://golang.org/ref/spec#Size_and_alignment_guarantees).

What reason were you thinking? Well-defined comparisons?
</blockquote></details>

Assigning `UnqualifiedStar{}` to an interface requires a runtime call that has small but measurable overhead. The difference is so minor that this isn't worth optimizing. I'm pretty sure this change doesn't change comparisons.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19160#-:-KwettBnJ0tjrZxBgkg9:b-48gj5d)*

**Comment by nvanbenschoten** [comment]
Yeah, I'd say no merge. Sorry for the misguided recommendation and thanks for putting this together.

**Comment by knz** [comment]
Thanks for the discussion!

**Pull request closed without merging**